### PR TITLE
API/OP（sequence_first_step，sequence_last_step，sequence_mask，beam_search，beam_search_decode） error message enhancement

### PR DESCRIPTION
**Pull request by jinyuKING**
API/OP（sequence_first_step，sequence_last_step，sequence_mask，beam_search，beam_search_decode） error message enhancement 
Both for python and C++ and add unit tests.

### CODE REVIEW

File: paddle/fluid/operators/beam_search_decode_op.cc [hunk 1]
```diff
@@ -129,13 +129,22 @@ class BeamSearchDecodeOp : public framework::OperatorBase {
     const LoDTensorArray* scores = ctx.Input<LoDTensorArray>("Scores");
     const size_t step_num = ids->size();
     PADDLE_ENFORCE_GT(step_num, 0UL,
-                      "beam search steps should be larger than 0");
+                      platform::errors::InvalidArgument(
+                          "beam search steps should be larger than"
```
Comment (guoshengCS) [inline paddle/fluid/operators/beam_search_decode_op.cc]: 建议通过Input()这种表述说明用户的输入有哪些问题，方便用户修改，如

`beam search steps, which is the size of Input(Ids) LoDTensorArray, ...`

File: paddle/fluid/operators/beam_search_decode_op.cc [hunk 2]
```diff
@@ -129,13 +129,22 @@ class BeamSearchDecodeOp : public framework::OperatorBase {
...
+                          "0, but received %d. ",
+                          step_num));
     const size_t source_num = ids->at(0).lod().at(0).size() - 1;
-    PADDLE_ENFORCE_GT(source_num, 0UL, "source num should be larger than 0");
+    PADDLE_ENFORCE_GT(source_num, 0UL, platform::errors::InvalidArgument(
+                                           "source_num should be larger than"
```
Comment (guoshengCS) [inline paddle/fluid/operators/beam_search_decode_op.cc]: source_num--> `sequence number of the first decoding step, indicating by Input(Ids)[0].lod[0].size, ...`

File: python/paddle/fluid/layers/rnn.py [hunk 3]
```diff
@@ -2993,6 +2993,12 @@ def beam_search(pre_ids,
                 beam_size=beam_size,
                 end_id=end_id)
     """
+    check_variable_and_dtype(pre_ids, 'pre_ids', ['int64'], 'beam_search')
+    check_variable_and_dtype(pre_scores, 'pre_scores', ['float32', 'float64'],
+                             'beam_search')
+    check_type(ids, 'ids', Variable, 'beam_search')
```
Comment (guoshengCS) [inline python/paddle/fluid/layers/rnn.py]: `ids`可以为None

File: paddle/fluid/operators/beam_search_decode_op.cc [hunk 4]
```diff
@@ -129,13 +129,22 @@ class BeamSearchDecodeOp : public framework::OperatorBase {
...
+                                           "0, but received %d. ",
+                                           source_num));
 
     for (size_t i = 0; i < step_num; ++i) {
-      PADDLE_ENFORCE_EQ(ids->at(i).lod().size(), 2UL,
-                        "Level of LodTensor should be 2");
+      PADDLE_ENFORCE_EQ(
+          ids->at(i).lod().size(), 2UL,
+          platform::errors::InvalidArgument("Level of LodTensor should be"
+                                            "2, but received %d. ",
```
Comment (guoshengCS) [inline paddle/fluid/operators/beam_search_decode_op.cc]: 可以如上进行调整

File: paddle/fluid/operators/beam_search_decode_op.h [hunk 5]
```diff
@@ -144,9 +149,15 @@ void BeamSearchDecoder<T>::Backtrace(const LoDTensorArray& step_ids,
                                      const LoDTensorArray& step_scores,
                                      LoDTensor* id_tensor,
                                      LoDTensor* score_tensor) const {
-  PADDLE_ENFORCE(!step_ids.empty(), "step num should be larger than 0");
+  PADDLE_ENFORCE_NE(
+      step_ids.empty(), true,
+      platform::errors::InvalidArgument("step num should be larger than 0."
```
Comment (guoshengCS) [inline paddle/fluid/operators/beam_search_decode_op.h]: `beam search steps, which is the size of Input(Ids) LoDTensorArray, ...`

File: paddle/fluid/operators/beam_search_decode_op.h [hunk 6]
```diff
@@ -144,9 +149,15 @@ void BeamSearchDecoder<T>::Backtrace(const LoDTensorArray& step_ids,
...
+                                        "But the input step_ids is empty."));
   PADDLE_ENFORCE_EQ(step_ids.size(), step_scores.size(),
-                    "step_ids and step_scores should be the same");
+                    platform::errors::InvalidArgument(
+                        "step_ids and step_scores should be the same."
```
Comment (guoshengCS) [inline paddle/fluid/operators/beam_search_decode_op.h]: 同上进行调整

File: paddle/fluid/operators/beam_search_decode_op.h [hunk 7]
```diff
@@ -82,7 +83,11 @@ void BeamSearchDecoder<T>::ConvertSentenceVectorToLodTensor(
     LoDTensor* score_tensor, bool reverse, bool sort_by_score) const {
   size_t src_num = sentence_vector_list.size();
 
-  PADDLE_ENFORCE_NE(src_num, 0, "src_num should not be 0");
+  PADDLE_ENFORCE_NE(src_num, 0,
+                    platform::errors::InvalidArgument("src_num has wrong value."
```
Comment (guoshengCS) [inline paddle/fluid/operators/beam_search_decode_op.h]: src_num --> `sequence number of the first decoding step, indicating by Input(Ids)[0].lod[0].size, ...`

File: paddle/fluid/operators/beam_search_op.h [hunk 8]
```diff
@@ -29,9 +29,15 @@ class BeamSearchOpKernel : public framework::OpKernel<T> {
     auto* pre_ids = context.Input<framework::LoDTensor>("pre_ids");
     auto* pre_scores = context.Input<framework::LoDTensor>("pre_scores");
 
-    PADDLE_ENFORCE_NOT_NULL(scores);
-    PADDLE_ENFORCE_NOT_NULL(pre_ids);
-    PADDLE_ENFORCE_NOT_NULL(pre_scores);
+    PADDLE_ENFORCE_NOT_NULL(scores,
+                            platform::errors::NotFound(
+                                "Input(score) of BeamSearchOp is not found."));
```
Comment (guoshengCS) [inline paddle/fluid/operators/beam_search_op.h]: Input(score)-->Input(scores)

File: paddle/fluid/operators/sequence_ops/sequence_mask_op.h [hunk 9]
```diff
@@ -80,7 +80,10 @@ class SequenceMaskKernel : public framework::OpKernel<Tx> {
     int maxlen = ctx.Attr<int>("maxlen");
     if (ctx.HasInput("MaxLenTensor")) {
       auto max_len_tensor = ctx.Input<Tensor>("MaxLenTensor");
-      PADDLE_ENFORCE(max_len_tensor != NULL, "MaxLenTensor is NULL");
+      PADDLE_ENFORCE_NOT_NULL(max_len_tensor,
+                              platform::errors::InvalidArgument(
+                                  "MaxLenTensor should not be NULL."
```
Comment (guoshengCS) [inline paddle/fluid/operators/sequence_ops/sequence_mask_op.h]: Input(MaxLenTensor)

File: paddle/fluid/operators/sequence_ops/sequence_mask_op.h [hunk 10]
```diff
@@ -94,7 +97,10 @@ class SequenceMaskKernel : public framework::OpKernel<Tx> {
       y->Resize(framework::make_ddim(y_dim));
 
       PADDLE_ENFORCE_GT(maxlen, 0,
-                        "MaxLenTensor value should be greater than 0");
+                        platform::errors::InvalidArgument(
+                            "MaxLenTensor value should be greater than 0. But "
```
Comment (guoshengCS) [inline paddle/fluid/operators/sequence_ops/sequence_mask_op.h]: `Input(MaxLenTensor) value should be greater than ...`

File: python/paddle/fluid/layers/sequence_lod.py [hunk 11]
```diff
@@ -539,6 +540,7 @@ def sequence_last_step(input):
              x = fluid.data(name='x', shape=[None, 10], dtype='float32', lod_level=1)
              x_last_step = fluid.layers.sequence_last_step(input=x)
     """
+    check_variable_and_dtype(input, 'input', ['float32'], 'sequence_last_step')
```
Comment (guoshengCS) [inline python/paddle/fluid/layers/sequence_lod.py]: Actually, `sequence_first_step `,  `sequence_last_step `, `sequence_pool ` support both float32 and float64. Maybe we can change these checks and docs

File: python/paddle/fluid/tests/unittests/sequence/test_sequence_mask.py [hunk 12]
```diff
@@ -154,5 +154,23 @@ def initParameters(self):
...
+
+            def test_Variable():
+                # the input must be Variable
+                fluid.layers.sequence_mask(input_data, maxlen=4)
+
+            self.assertRaises(TypeError, test_Variable)
+
+            def test_input_dtype():
+                # the dtype of input must be int64
+                fluid.layers.sequence_mask(input_data, maxlen=4)
```
Comment (guoshengCS) [inline python/paddle/fluid/tests/unittests/sequence/test_sequence_mask.py]: 与上一个单测重复, 是否是写错了

File: python/paddle/fluid/tests/unittests/test_beam_search_op.py [hunk 13]
```diff
@@ -96,5 +98,109 @@ def _create_scores(self):
         tensor.set_lod(self.lod)
 
 
+class TestBeamSearchOpError(unittest.TestCase):
+    def test_errors(self):
+        with program_guard(Program(), Program()):
+            pre_ids = fluid.layers.data(
```
Comment (guoshengCS) [inline python/paddle/fluid/tests/unittests/test_beam_search_op.py]: 以后最好使用新的`fluid.data` api

File: paddle/fluid/operators/beam_search_decode_op.cc (hunk 1 above)
Comment (jinyuKING) [reply paddle/fluid/operators/beam_search_decode_op.cc]: > 建议通过Input()这种表述说明用户的输入有哪些问题，方便用户修改，如
> 
> `beam search steps, which is the size of Input(Ids) LoDTensorArray, ...`

done. thanks.

File: paddle/fluid/operators/beam_search_decode_op.cc (hunk 2 above)
Comment (jinyuKING) [reply paddle/fluid/operators/beam_search_decode_op.cc]: > source_num--> `sequence number of the first decoding step, indicating by Input(Ids)[0].lod[0].size, ...`

done. thanks

File: python/paddle/fluid/layers/rnn.py (hunk 3 above)
Comment (jinyuKING) [reply python/paddle/fluid/layers/rnn.py]: > `ids`可以为None

done. thanks

File: paddle/fluid/operators/beam_search_decode_op.cc (hunk 4 above)
Comment (jinyuKING) [reply paddle/fluid/operators/beam_search_decode_op.cc]: > 可以如上进行调整

done. thanks

File: paddle/fluid/operators/beam_search_decode_op.h (hunk 5 above)
Comment (jinyuKING) [reply paddle/fluid/operators/beam_search_decode_op.h]: > `beam search steps, which is the size of Input(Ids) LoDTensorArray, ...`

done. thanks

File: paddle/fluid/operators/beam_search_decode_op.h (hunk 6 above)
Comment (jinyuKING) [reply paddle/fluid/operators/beam_search_decode_op.h]: > 同上进行调整

done. thanks.

File: paddle/fluid/operators/beam_search_decode_op.h (hunk 7 above)
Comment (jinyuKING) [reply paddle/fluid/operators/beam_search_decode_op.h]: > src_num --> `sequence number of the first decoding step, indicating by Input(Ids)[0].lod[0].size, ...`

done. thanks

File: paddle/fluid/operators/beam_search_op.h (hunk 8 above)
Comment (jinyuKING) [reply paddle/fluid/operators/beam_search_op.h]: > Input(score)-->Input(scores)

done. thanks

File: paddle/fluid/operators/sequence_ops/sequence_mask_op.h (hunk 9 above)
Comment (jinyuKING) [reply paddle/fluid/operators/sequence_ops/sequence_mask_op.h]: > Input(MaxLenTensor)

done. thanks

File: paddle/fluid/operators/sequence_ops/sequence_mask_op.h (hunk 10 above)
Comment (jinyuKING) [reply paddle/fluid/operators/sequence_ops/sequence_mask_op.h]: > `Input(MaxLenTensor) value should be greater than ...`

done. thanks.

File: python/paddle/fluid/layers/sequence_lod.py (hunk 11 above)
Comment (jinyuKING) [reply python/paddle/fluid/layers/sequence_lod.py]: > Actually, `sequence_first_step `, `sequence_last_step `, `sequence_pool ` support both float32 and float64. Maybe we can change these checks and docs

done. thanks

File: python/paddle/fluid/tests/unittests/sequence/test_sequence_mask.py (hunk 12 above)
Comment (jinyuKING) [reply python/paddle/fluid/tests/unittests/sequence/test_sequence_mask.py]: > 与上一个单测重复, 是否是写错了

done.thanks

File: python/paddle/fluid/tests/unittests/test_beam_search_op.py (hunk 13 above)
Comment (jinyuKING) [reply python/paddle/fluid/tests/unittests/test_beam_search_op.py]: > 以后最好使用新的`fluid.data` api

done.thanks

**Review by guoshengCS — APPROVED** [review]
LGTM

**Comment by guoshengCS** [comment]
image

部分文档有调整，还请 @swtkiwi 帮忙review下

**Review by swtkiwi — APPROVED** [review]
LGTM

**Pull request merged**